node/core/approval-voting/src/lib.rs

```diff
@@ -539,6 +533,95 @@ struct ApprovalStatus {
 	block_tick: Tick,
 }
 
+enum ApprovalOutcome {
+	Approved,
+	Failed(InvalidCandidate),
+}
+
+struct ApprovalState {
+	validator_index: ValidatorIndex,
+	candidate_hash: CandidateHash,
+	candidate_index: CandidateIndex,
+	status: ApprovalOutcome,
+}
+
+impl ApprovalState {
+	fn approved(
+		validator_index: ValidatorIndex,
+		candidate_hash: CandidateHash,
+		candidate_index: CandidateIndex
+	) -> Self {
+		Self {
+			validator_index,
+			candidate_hash,
+			candidate_index,
+			status: ApprovalOutcome::Approved
+		}
+	}
+	fn failed(
+		validator_index: ValidatorIndex,
+		candidate_hash: CandidateHash,
+		candidate_index: CandidateIndex,
+		reason: InvalidCandidate
+	) -> Self {
+		Self {
+			validator_index,
+			candidate_hash,
+			candidate_index,
+			status: ApprovalOutcome::Failed(reason)
+		}
+	}
+}
+
+type BoundedBackgroundApprovalTask = Timeout<RemoteHandle<ApprovalState>>;
+
+struct CurrentlyCheckingSet {
+	approvals_cache: lru::LruCache<CandidateHash, ApprovalOutcome>,
+	candidate_hash_map: HashMap<CandidateHash, Vec<Hash>>,
+	currently_checking: FuturesUnordered<BoundedBackgroundApprovalTask>,
+}
+
+impl Default for CurrentlyCheckingSet {
+	fn default() -> Self {
+		Self {
+			approvals_cache: lru::LruCache::new(APPROVAL_CACHE_SIZE),
+			candidate_hash_map: HashMap::new(),
+			currently_checking: FuturesUnordered::new(),
+		}
+	}
+}
+
+impl CurrentlyCheckingSet {
+	fn insert_relay_block_hash(
+		&mut self,
+		candidate_hash: &CandidateHash,
+		relay_block: Hash,
+	) -> bool {
```

Comment:
This differs from the API I proposed, which combined the "contains" check with the conditional insertion if contains is false.

That's fine, but the `CurrentlyCheckingSet` should also get an `insert_work` function so users don't reach into its private internals.

It's unclear from the function signature whether the bool here indicates whether the candidate or the block was unknown. A comment on the function should